Dependencies/patches/libass-unicode-fontdir-path-win.patch

```diff
@@ -0,0 +1,94 @@
+--- a/libass/ass_fontselect.c	2017-10-27 01:34:18.000000000 +0800
++++ b/libass/ass_fontselect.c	2020-05-04 01:50:49.144060600 +0800
+@@ -47,6 +47,9 @@
+ #include "ass_font.h"
+ #include "ass_string.h"
+
++#include <stringapiset.h>
++#include <windows.h>
++
+ #define ABS(x) ((x) < 0 ? -(x) : (x))
+ #define MAX_FULLNAME 100
+
+@@ -163,26 +166,62 @@ static ASS_FontProviderFuncs ft_funcs =
+
+ static void load_fonts_from_dir(ASS_Library *library, const char *dir)
+ {
+-    DIR *d = opendir(dir);
+-    if (!d)
+-        return;
+-    while (1) {
+-        struct dirent *entry = readdir(d);
+-        if (!entry)
+-            break;
+-        if (entry->d_name[0] == '.')
+-            continue;
+-        char fullname[4096];
+-        snprintf(fullname, sizeof(fullname), "%s/%s", dir, entry->d_name);
+-        size_t bufsize = 0;
+-        ass_msg(library, MSGL_INFO, "Loading font file '%s'", fullname);
+-        void *data = read_file(library, fullname, &bufsize);
+-        if (data) {
+-            ass_add_font(library, entry->d_name, data, bufsize);
+-            free(data);
+-        }
++    wchar_t wpath[PATH_MAX];
++    if (!MultiByteToWideChar(CP_UTF8, 0, dir, -1, wpath, PATH_MAX)) {
++        ass_msg(library, MSGL_WARN,
++                "Error obtaining unicode path for font dir: '%s' (code: %d)", dir, GetLastError());
++        return ;
++    }
++    wchar_t wpath_wc[PATH_MAX];
++    wcscpy(wpath_wc,wpath);
++    wcscat(wpath_wc, L"\\*.*");
++    HANDLE hFind;
++    WIN32_FIND_DATAW FindFileData;
++    if((hFind = FindFirstFileW(wpath_wc, &FindFileData)) != INVALID_HANDLE_VALUE){
++        do{
++            if(!(FindFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
++                HANDLE hFile;
++                wchar_t fullName[PATH_MAX];
++                wcscpy(fullName,wpath);
++                wcscat(fullName, L"\\");
++                wcscat(fullName, FindFileData.cFileName);
```

Comment:
This and above can overflow too.

Reply:
But a filename's maximum full path length is `PATH_MAX`, I don't think it can overflow here.

Reply:
maybe using a `std::wstring` is better here?

Reply:
It's C, so you cannot use C++. `PATH_MAX` on Windows is rather small, and the path there can actually be longer. You can use `wcscat_s` probably.